File: python/stack/binary-tree-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        if root is None:
            return []
        return self.postorderTraversal(root.left) +\
            self.postorderTraversal(root.right) + [root.val]

File: go/stack/binary-tree-postorder-traversal.go
package stack

type TreeNode struct {
     Val int
     Left *TreeNode
     Right *TreeNode
}

func postorderTraversal(root *TreeNode) []int {
	var result []int

	if root == nil {
		return result
	}

	st := New()
	st.Push(root)
	for  !st.IsEmpty() {
		tmp := st.Pop().(*TreeNode)
		result = append(result, tmp.Val)

		cur := tmp.Left
		if cur != nil {
			st.Push(cur)
			cur = cur.Left
		}

		cur = tmp.Right
		if cur != nil {
			st.Push(cur)
			cur = cur.Right
		}
	}
	i, j := 0, len(result)-1
	for i < j {
		result[i], result[j] = result[j], result[i]
		i++
		j--
	}
	return result
}
